lib/instrumentation/modules/@aws-sdk/client-sns.js

```diff
@@ -0,0 +1,198 @@
+/*
+ * Copyright Elasticsearch B.V. and other contributors where applicable.
+ * Licensed under the BSD 2-Clause License; you may not use this file except in
+ * compliance with the BSD 2-Clause License.
+ */
+
+'use strict';
+
+const constants = require('../../../constants');
+const NAME = 'SNS';
+const TYPE = 'messaging';
+const SUBTYPE = 'sns';
+const MAX_SNS_MESSAGE_ATTRIBUTES = 10;
+const elasticAPMStash = Symbol('elasticAPMStash');
+
+/**
+ * Returns middlewares to instrument an S3Client instance
+ *
+ * @param {import('@aws-sdk/client-sns').SNSClient} client
+ * @param {any} agent
+ * @returns {import('./smithy-client').AWSMiddlewareEntry[]}
+ */
+function snsMiddlewareFactory(client, agent) {
+  return [
+    {
+      middleware: (next, context) => async (args) => {
+        const ins = agent._instrumentation;
+        const log = agent.logger;
+        const span = ins.currSpan();
+        const input = args.input;
+
+        // W3C trace-context propagation.
+        const runContext = ins.currRunContext();
+        const parentSpan =
+          span || runContext.currSpan() || runContext.currTransaction();
+        const targetId =
+          input && (input.TopicArn || input.TargetArn || input.PhoneNumber);
+
+        // Though our spec only mentions a 10-message-attribute limit for *SQS*, we'll
+        // do the same limit here, because
+        // https://docs.aws.amazon.com/sns/latest/dg/sns-message-attributes.html
+        // mentions the 10-message-attribute limit for SQS subscriptions.
+        input.MessageAttributes = input.MessageAttributes || {};
+        if (
+          input.MessageAttributes &&
+          Object.keys(input.MessageAttributes).length + 2 >
+            MAX_SNS_MESSAGE_ATTRIBUTES
+        ) {
+          log.warn(
+            'cannot propagate trace-context with SNS message to %s, too many MessageAttributes',
+            targetId,
+          );
+        } else {
+          parentSpan.propagateTraceContextHeaders(
+            input.MessageAttributes,
+            function (msgAttrs, name, value) {
+              if (name.startsWith('elastic-')) {
+                return;
+              }
+              msgAttrs[name] = { DataType: 'String', StringValue: value };
+            },
+          );
+        }
+
+        // Ensure there is a span from the wrapped `client.send()`.
+        if (!span || !(span.type === TYPE && span.subtype === SUBTYPE)) {
+          return await next(args);
+        }
+
+        let queueName;
+
+        if (input.PhoneNumber) {
```

Comment:
(I think this is minor, because I'm not sure it is possible to have an `input` that has more than one of `PhoneNumber`, `TopicArn`, or `TargetArn`.)

In this code block, `PhoneNumber` "wins" for setting queueName. In the `shouldIgnore...()` function below, `TopicArn` "wins".  I see that this same inconsistency exists in "aws-sdk/sns.js". Do you know if SNS docs provide a hint here: Can there be more than one of these vars set? If so, should one take precedence?

Reply:
Honestly I haven't thought about combinations of props. Thanks to pointing out :)

If I recall properly the docs `input` is just an object containing all options/params of the command. Typescript yells at you if you pass a property that is not defined in the command but you can always be nasty and use the type `any` to bypass the compiler. What I'm not sure is if there is a runtime check for the input object.

Reply:
We can discuss after I finish my findings in
https://github.com/elastic/apm-agent-nodejs/pull/3532#discussion_r1329926158